### ISSUE TITLE
Build error (58.0.3029.110/ develop branch) on Archlinux

### ISSUE THREAD

**Issue by Alex2242** [Eloston/ungoogled-chromium#266]
I tried to build the latest release (58.0.3029.110) on Archlinux but compilation failed :

```
[347/347] c++ -pthread -o gn -Wl,--start-group tools/gn/gn_main.o libevent.a base.a xdg_user_dirs.a gn_lib.a dynamic_annotations.a -Wl,--end-group -lrt -latomic
FAILED: gn 
c++ -pthread -o gn -Wl,--start-group tools/gn/gn_main.o libevent.a base.a xdg_user_dirs.a gn_lib.a dynamic_annotations.a -Wl,--end-group -lrt -latomic
base/threading/thread_task_runner_handle.o: In function 'base::ThreadTaskRunnerHandle::OverrideForTesting(scoped_refptr<base::SingleThreadTaskRunner>)':
thread_task_runner_handle.cc:(.text+0x10a5): undefined reference to 'base::ScopedClosureRunner::ScopedClosureRunner(base::Callback<void (), (base::internal::CopyMode)1, (base::internal::RepeatMode)1> const&)'
thread_task_runner_handle.cc:(.text+0x1168): undefined reference to 'base::ScopedClosureRunner::ScopedClosureRunner(base::Callback<void (), (base::internal::CopyMode)1, (base::internal::RepeatMode)1> const&)'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Command '['ninja', '-C', '/tmp/tmpZwENSm', '-v', 'gn']' returned non-zero exit status 1
```

I could not build develop branch either :

```
[367/367] c++ -pthread -o gn -Wl,--start-group tools/gn/gn_main.o libevent.a base.a xdg_user_dirs.a gn_lib.a dynamic_annotations.a -Wl,--end-group -lrt -latomic
FAILED: gn 
c++ -pthread -o gn -Wl,--start-group tools/gn/gn_main.o libevent.a base.a xdg_user_dirs.a gn_lib.a dynamic_annotations.a -Wl,--end-group -lrt -latomic
base/task_scheduler/task_scheduler_impl.o: In function 'base::internal::TaskSchedulerImpl::CreateTaskRunnerWithTraits(base::TaskTraits const&)':
task_scheduler_impl.cc:(.text+0x1f5): undefined reference to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)'
base/task_scheduler/task_scheduler_impl.o: In function 'base::internal::TaskSchedulerImpl::CreateSequencedTaskRunnerWithTraits(base::TaskTraits const&)':
task_scheduler_impl.cc:(.text+0x255): undefined reference to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)'
base/task_scheduler/task_scheduler_impl.o: In function 'base::internal::TaskSchedulerImpl::GetMaxConcurrentTasksWithTraitsDeprecated(base::TaskTraits const&) const':
task_scheduler_impl.cc:(.text+0x298): undefined reference to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)'
base/task_scheduler/task_scheduler_impl.o: In function 'base::internal::TaskSchedulerImpl::PostDelayedTaskWithTraits(tracked_objects::Location const&, base::TaskTraits const&, base::Callback<void (), (base::internal::CopyMode)0, (base::internal::RepeatMode)0>, base::TimeDelta)':
task_scheduler_impl.cc:(.text+0x2e3): undefined reference to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)'
base/task_scheduler/task_scheduler_impl.o: In function 'base::internal::TaskSchedulerImpl::GetWorkerPoolForTraits(base::TaskTraits const&) const':
task_scheduler_impl.cc:(.text+0x968): undefined reference to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)'
base/task_scheduler/scheduler_single_thread_task_runner_manager.o:scheduler_single_thread_task_runner_manager.cc:(.text+0x1287): more undefined references to 'base::internal::GetEnvironmentIndexForTraits(base::TaskTraits const&)' follow
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Command '['ninja', '-C', '/tmp/tmpcBxbHa', '-v', 'gn']' returned non-zero exit status 1
```
Am I missing something ?

**Comment by Eloston** [comment]
I'm not sure at this point. How are you building it?

**Comment by Alex2242** [comment]
I tried to follow BUILDING.md   : 
```
export UTILIKIT_CONFIG_TYPE=linux_simple
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
./utilikit/generate_build_files.py linux_simple --apply-domain-substitution
./ungoogled_linux_simple/build.sh
```

**Comment by Eloston** [comment]
`linux_simple` is not a valid utilikit config type. Perhaps you mean `linux_conservative`? You could also try `archlinux`.

The other commands look fine to me.

**Comment by Alex2242** [comment]
The reason I put `linux_simple` is this : 
```
$ ./utilikit/generate_build_files.py archlinux --apply-domain-substitution
generate_build_files.py: error: argument files_type: invalid choice: 'archlinux' (choose from 'debian', 'macos', 'linux_simple')
```

**Comment by Eloston** [comment]
The documentation (as it is now) does not explain it clearly, but the `utilikit` config type and the build files type are not the same thing. The `utilikit` config type is set through an environment variable, and the build files type is set in the argument to `generate_build_files.py`. Only the `utilikit` config type needs to be changed.

**Comment by avently** [comment]
@Alex2242 did you successfully build ungoogled on arch?
I had build error related to ICU (no rules to build a target). Fixed it. But build stopped on error related to Breakpad.

**Comment by Alex2242** [comment]
@Eloston  Thanks, my bad.
@avently No, it still fails. I think I've the same error you had :
``` 
ninja: error: '../../components/domain_reliability/baked_in_configs/c_bigcache_googleapis_com.json', needed by 'obj/components/domain_reliability/bake_in_configs.inputdeps.stamp', missing and no known rule to make it
```
 but I haven't found a way to fix it. How did you do ?

**Comment by avently** [comment]
@Alex2242 I have no time and hardware resources to try to fix every error I get while building software. I think the developer can make as easy build configuration as possible. One hour of developer time can save thousands hours of time for regular users that need to fix build errors. 
I don't know why build fails on my arch. Maybe gn attributes is wrong. Removing gn attributes one by one can help.

**Comment by Eloston** [comment]
@Alex2242 Are you trying to build the `develop` branch? You shouldn't be getting that error on `master`.

**Comment by avently** [comment]
@Eloston this happens on master.
export UTILIKIT_CONFIG_TYPE=archlinux
mkdir build/                                                                                                                                                                                       
mkdir build/sandbox
mkdir build/downloads
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
./utilikit/generate_build_files.py linux_simple --apply-domain-substitution                                                                                                                        
cd build/sandbox
ungoogled_linux_simple/build.sh
```
ninja: Entering directory `out/Default'
ninja: error: '../../third_party/icu/common/icudtl.dat', needed by 'obj/third_party/icu/make_data_assembly.inputdeps.stamp', missing and no known rule to make it
```

touch third_party/icu/common/icudtl.dat
ninja -C out/Default chrome

```
ninja: Entering directory `out/Default'
[47/26165] CXX obj/breakpad/client/exception_handler.o
FAILED: obj/breakpad/client/exception_handler.o 
../../../../../../../../usr/bin/clang++ -MMD -MF obj/breakpad/client/exception_handler.o.d -DV8_DEPRECATION_WARNINGS -DUSE_UDEV -DUI_COMPOSITOR_IMAGE_TRANSPORT -DUSE_AURA=1 -DUSE_PANGO=1 -DUSE_CAIRO=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_X11=1 -DNO_TCMALLOC -DDISABLE_NACL -DCHROMIUM_BUILD -DENABLE_MEDIA_ROUTER=1 -DCR_CLANG_REVISION=\"296320-1\" -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -I../../breakpad -I../../breakpad/src -I../../breakpad/src/client -I../../breakpad/src/third_party/linux/include -I../.. -Igen -I../../breakpad/src -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -funwind-tables -fPIC -pipe -fcolor-diagnostics -m64 -march=x86-64 -pthread -O2 -fno-ident -fdata-sections -ffunction-sections -g0 -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wall -Wno-unused-variable -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-covered-switch-default -Wno-deprecated-register -Wno-unneeded-internal-declaration -Wno-inconsistent-missing-override -Wno-shift-negative-value -Wno-undefined-var-template -Wno-nonportable-include-path -Wno-address-of-packed-member -Wno-block-capture-autoreleasing -Wno-unused-lambda-capture -Wno-user-defined-warnings -fvisibility-inlines-hidden -std=gnu++11 -fno-rtti -fno-exceptions -Wno-deprecated -c ../../breakpad/src/client/linux/handler/exception_handler.cc -o obj/breakpad/client/exception_handler.o
In file included from ../../breakpad/src/client/linux/handler/exception_handler.cc:66:
../../breakpad/src/client/linux/handler/exception_handler.h:194:21: error: field has incomplete type 'struct ucontext'
    struct ucontext context;
                    ^
../../breakpad/src/client/linux/handler/exception_handler.h:194:12: note: forward declaration of 'google_breakpad::ucontext'
    struct ucontext context;
           ^
../../breakpad/src/client/linux/handler/exception_handler.cc:442:41: error: invalid application of 'sizeof' to an incomplete type 'struct ucontext'
  memcpy(&g_crash_context_.context, uc, sizeof(struct ucontext));
                                        ^     ~~~~~~~~~~~~~~~~~
../../breakpad/src/client/linux/handler/exception_handler.h:194:12: note: forward declaration of 'google_breakpad::ucontext'
    struct ucontext context;
           ^
../../breakpad/src/client/linux/handler/exception_handler.cc:456:13: error: member access into incomplete type 'struct ucontext'
  if (uc_ptr->uc_mcontext.fpregs) {
            ^
../../breakpad/src/client/linux/handler/exception_handler.h:194:12: note: forward declaration of 'google_breakpad::ucontext'
    struct ucontext context;
           ^
../../breakpad/src/client/linux/handler/exception_handler.cc:457:49: error: member access into incomplete type 'struct ucontext'
    memcpy(&g_crash_context_.float_state, uc_ptr->uc_mcontext.fpregs,
                                                ^
../../breakpad/src/client/linux/handler/exception_handler.h:194:12: note: forward declaration of 'google_breakpad::ucontext'
    struct ucontext context;
           ^
../../breakpad/src/client/linux/handler/exception_handler.cc:479:19: error: variable has incomplete type 'struct ucontext'
  struct ucontext context;
                  ^
../../breakpad/src/client/linux/handler/exception_handler.h:194:12: note: forward declaration of 'google_breakpad::ucontext'
    struct ucontext context;
           ^
5 errors generated.
[52/26165] CXX obj/base/base/file_util_posix.o
ninja: build stopped: subcommand failed.
```

**Comment by Eloston** [comment]
@avently This seems to be a problem with ungoogled-chromium's code, but I'm not sure what the problem could be. I think the problems with the other configurations show that Archlinux support is finicky at best.

For now, I think it's best to hold off until the next stable version of ungoogled-chromium. I think trying to fix 58 may be more effort than it's worth.

**Comment by xsmile** [comment]
@avently Most probably you need to apply breakpad-use-ucontext_t.patch from the official Arch Linux chromium package, see https://git.archlinux.org/svntogit/packages.git/tree/trunk?h=packages/chromium.

**Comment by avently** [comment]
@xsmile yes, now it builds fine. Thank you!
@Eloston please, include this patch into your repo. In arch repo chromium version is 62 and this patch will be compatable with your next milestone.

**Comment by Eloston** [comment]
@avently Alright, will do :+1:

**Comment by avently** [comment]
@Eloston also, what about
```
ninja: error: '../../third_party/icu/common/icudtl.dat', needed by 'obj/third_party/icu/make_data_assembly.inputdeps.stamp', missing and no known rule to make it
```
?
touch third_party/icu/common/icudtl.dat   -- doesn't work. Later i received an error related to this file. I could only return this file from chromium repo because it gets deleted via cleaning_list.

**Comment by Eloston** [comment]
@avently The GN flag `icu_use_data_file=false` is supposed to disable the use of `icudtl.dat`, and it works on Debian. Debian and Arch essentially have the same GN configuration, so I don't know why this is failing specifically on Arch.

**Comment by Eloston** [comment]
After a quick search, it seems the only thing that might be making a difference is the `unbundle` script. This script essentially replaces some GN files so Chromium builds against more systme libraries. Since [the ICU library has a corresponding file](https://chromium.googlesource.com/chromium/src/+/58.0.3029.110/build/linux/unbundle/icu.gn), the unbundle script is probably causing Chromium to use the system ICU library, which may be bypassing the dependency on `icudtl.dat` (and also `icu_use_data_file=false`).

So, you could either:

* Remove `icu_use_data_file=false` and add back the original `icudtl.dat`
* Run the `unbundle` script in `resources/packaging/debian/standard/scripts/`, which will cause Chromium to build against more system libraries. But I have a feeling that this may break something else since 58 is pretty old.

**Comment by avently** [comment]
@Eloston for now i'm building with icu. In the next version i'll try second option

**Comment by avently** [comment]
@Eloston another build error
```
LINK ./font_service.service
FAILED: font_service.service 
../../../../../../../../usr/bin/clang++ -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -Wl,--no-as-needed -lpthread -Wl,--as-needed -fuse-ld=gold -B -Wl,--icf=all -m64 -pthread -Wl,-O1 -Wl,--gc-sections -Wl,-rpath-link=. -Wl,--disable-new-dtags -Wl,--export-dynamic -o "./font_service.service" -Wl,--start-group @"./font_service.service.rsp"  -Wl,--end-group  -ldl -lrt -lgmodule-2.0 -lgobject-2.0 -lgthread-2.0 -lglib-2.0 -latomic -lsmime3 -lnss3 -lnssutil3 -lplds4 -lplc4 -lnspr4 -lpangocairo-1.0 -lpango-1.0 -lcairo -lfontconfig -lfreetype -lexpat -lgio-2.0 
obj/base/libbase.a(message_pump_libevent.o):../../base/message_loop/message_pump_libevent.cc:function base::MessagePumpLibevent::WatchFileDescriptor(int, bool, int, base::MessagePumpLibevent::FileDescriptorWatcher*, base::MessagePumpLibevent::Watcher*): error: undefined reference to 'event_get_fd'
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
[8/11380] CXX obj/media/capture/mojo/...ure_types/video_capture_types.mojom.o
ninja: build stopped: subcommand failed.
```
Seems like this line ruins the build:
https://github.com/Eloston/ungoogled-chromium/blob/d3f382b9c723aea48b1a50ed533b303c91d02498/resources/patches/debian/system/event.patch#L82
I returned the old line and then the build continued.

**Comment by avently** [comment]
another error:
```
FAILED: obj/third_party/WebKit/Source/platform/platform/SharedGpuContext.o 
../../../../../../../../usr/bin/clang++ -MMD -MF obj/third_party/WebKit/Source/platform/platform/SharedGpuContext.o.d -DBLINK_PLATFORM_IMPLEMENTATION=1 -DINSIDE_BLINK -DV8_DEPRECATION_WARNINGS -DUSE_UDEV -DUI_COMPOSITOR_IMAGE_TRANSPORT -DUSE_AURA=1 -DUSE_PANGO=1 -DUSE_CAIRO=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_X11=1 -DNO_TCMALLOC -DDISABLE_NACL -DCHROMIUM_BUILD -DENABLE_MEDIA_ROUTER=1 -DCR_CLANG_REVISION=\"296320-1\" -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_32 -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_26 -DGL_GLEXT_PROTOTYPES -DUSE_GLX -DUSE_EGL -DSK_IGNORE_DW_GRAY_FIX -DSK_IGNORE_DIRECTWRITE_GASP_FIX -DSK_IGNORE_LINEONLY_AA_CONVEX_PATH_OPTS -DSK_SUPPORT_GPU=1 -DGOOGLE_PROTOBUF_NO_RTTI -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER -DHAVE_PTHREAD -DENABLE_LAYOUT_UNIT_IN_INLINE_BOXES=0 -DENABLE_OILPAN=1 -DWTF_USE_CONCATENATED_IMPULSE_RESPONSES=1 -DWTF_USE_WEBAUDIO_FFMPEG=1 -DWTF_USE_DEFAULT_RENDER_THEME=1 -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DU_STATIC_IMPLEMENTATION -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_STATIC -DUSE_SYSTEM_LIBJPEG -DENABLE_IPC_FUZZER -Igen/blink -I../../third_party/ffmpeg -I../.. -Igen -I../../third_party/WebKit/Source -I../../third_party/WebKit -Igen/blink -Igen/third_party/WebKit -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I../../third_party/khronos -I../../gpu -I../../third_party/libwebp -I../../skia/config -I../../skia/ext -I../../third_party/skia/include/c -I../../third_party/skia/include/config -I../../third_party/skia/include/core -I../../third_party/skia/include/effects -I../../third_party/skia/include/images -I../../third_party/skia/include/lazy -I../../third_party/skia/include/pathops -I../../third_party/skia/include/pdf -I../../third_party/skia/include/pipe -I../../third_party/skia/include/ports -I../../third_party/skia/include/utils -I../../third_party/skia/include/gpu -I../../third_party/skia/src/gpu -I../../third_party/skia/src/sksl -I../../third_party/libwebm/source -I../../third_party/protobuf/src -Igen/protoc_out -I../../third_party/protobuf/src -I../../third_party/boringssl/src/include -I/usr/include/nss -I/usr/include/nspr -I../../third_party/icu/source/common -I../../third_party/icu/source/i18n -I../../third_party/iccjpeg -I../../third_party/libpng -I../../third_party/zlib -I../../third_party/ots/include -I../../v8/include -Igen/v8/include -I../../third_party/ced/src -I../../third_party/harfbuzz-ng/src -I../../third_party/ffmpeg/chromium/config/ChromeOS/linux/x64 -I../../third_party/ffmpeg -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -funwind-tables -fPIC -pipe -fcolor-diagnostics -m64 -march=x86-64 -pthread -Wall -Wextra -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-covered-switch-default -Wno-deprecated-register -Wno-unneeded-internal-declaration -Wno-inconsistent-missing-override -Wno-shift-negative-value -Wno-undefined-var-template -Wno-nonportable-include-path -Wno-address-of-packed-member -Wno-block-capture-autoreleasing -Wno-unused-lambda-capture -Wno-user-defined-warnings -O2 -fno-ident -fdata-sections -ffunction-sections -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wglobal-constructors -g0 -Wno-header-guard -fvisibility-inlines-hidden -std=gnu++11 -fno-rtti -fno-exceptions -c ../../third_party/WebKit/Source/platform/graphics/gpu/SharedGpuContext.cpp -o obj/third_party/WebKit/Source/platform/platform/SharedGpuContext.o
In file included from ../../third_party/WebKit/Source/platform/graphics/gpu/SharedGpuContext.cpp:5:
../../third_party/WebKit/Source/platform/graphics/gpu/SharedGpuContext.h:38:16: error: no template named 'function' in namespace 'std'
  typedef std::function<std::unique_ptr<WebGraphicsContext3DProvider>()>
          ~~~~~^
../../third_party/WebKit/Source/platform/graphics/gpu/SharedGpuContext.h:53:53: error: cannot initialize a member subobject of type 'blink::SharedGpuContext::ContextProviderFactory' (aka 'int') with an rvalue of type 'nullptr_t'
  ContextProviderFactory m_contextProviderFactory = nullptr;
                                                    ^~~~~~~
2 errors generated.
```
fixed by suggest from https://bugs.webkit.org/show_bug.cgi?id=162268
Solution was to add 
#include <functional> 
to
third_party/WebKit/Source/platform/graphics/gpu/SharedGpuContext.h

Than build was successfull

**Comment by Eloston** [comment]
Fixed in 5ca1b6e7bdcad5d32b5a998c69312758f681e6c1
